### ISSUE TITLE
Non-nul negative_approx_kl (=log(ratio)) at first ppo_epoch

### ISSUE THREAD

**Issue by dorian-bischoff** [volcengine/verl#2305]
Hello,

I am currently working with verl and observed the following unexpected behavior. When updating the parameters of the model for the first minibatch in the first ppo epoch, using GRPO, we have $\pi=\pi_{old}$, thus `negative_approx_kl` (from `compute_policy_loss` in verl/trainers/ppo/core_algos.py) should be 0. But this is not always the case. I observe that most of them are nul, but sometimes it is not the case for some sequences.

What I observe:
- log_prob!=old_log_prob for one or more sequences and for one or several ranks.
- In the same minibatch, if a given rank had log_prob!=old_log_prob in the first microbatch, it is always the case for the other microbatches of the given minibatch. The next minibatch is independent: the same given rank can have log_prob==old_log_prob or not.
- There is no clear increase or decrease of the number of ranks concerned by this issue during training (could be for example no rank, then rank 3 then ranks 2, 5 and 6 then no rank, etc...).
- When log_prob!=old_log_prob, this is only for one or some sequences, not all the sequences of the microbatch (at least I have not observed the whole microbatch being concerned). I generally found that it was the sequences at the end of the microbatch (but not always).
- I have checked that for a given sequence, if log_prob[seq_idx]!=old_log_prob[seq_idx], the log_prob and old_log_prob have been computed using the same input_ids and attention_mask in `compute_log_prob` function and `update_policy` function (and no problem there).
- If I recompute the old_log_prob just before the update step (e.g. before this [line](https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/ray_trainer.py#L1266)), old_log_prob==old_log_prob_recomputed. Same observation when I recompute the log_prob for a microbatch (e.g. before this [line](https://github.com/volcengine/verl/blob/main/verl/workers/actor/dp_actor.py#L539)), log_prob==log_prob_recomputed. And we still have log_prob!=old_log_prob and log_prob_recomputed!=old_log_prob_recomputed
- I have tried the following configurations: 1, 2, 4 or 8 H200 or 4 H100, all ended up in the same observation of log_prob/old_log_prob discrepancies.
- I have tried using vllm or sglang -> same observation of log_prob/old_log_prob discrepancies.
- I have simplified the reproduction (see below) with TinyZero, but I was originately working with SearchR1, so I believe a dataset influence or multiturn issue can be excluded.

In the provided log, we have the rank 7 which is non-zero for the first ppo_epoch and first minibatch (all other ranks are concerned in my other tests). [example_log.log](https://github.com/user-attachments/files/20999720/example_log.log)

How to reproduce:
I use a TinyZero-like setup to run my tests:
- Create folder and environment
```
mkdir TinyZero
cd TinyZero
git clone https://github.com/volcengine/verl.git ./

conda create -n zero python==3.10 -y
conda activate zero
pip install -e .[vllm]
pip3 install flash-attn==2.7.4.post1 --no-build-isolation
```
- For the dataset, copy the dataset preprocessing [file](https://github.com/Jiayi-Pan/TinyZero/blob/main/examples/data_preprocess/countdown.py) in the folder TinyZero, then
```
mkdir /data/tinyzero
python TinyZero/data_preprocess_countdown.py --local_dir /data/tinyzero --template_type qwen-instruct
```
- Copy the reward function [file](https://github.com/Jiayi-Pan/TinyZero/blob/main/verl/utils/reward_score/countdown.py) in the folder TinyZero/verl/utils/reward_score
- To print discrepancies, in the file /TinyZero/verl/workers/actor/dp_actors.py:
    - Change line 446 `for data in micro_batches:` -> `for micro_batch_idx, data in enumerate(micro_batches):`
    - Add line 495 `print(f"[RANK {torch.distributed.get_rank()}] -- PPO Epoch {epoch+1} | Minibatch {batch_idx+1} | Microbatch {micro_batch_idx+1} | Summed difference (log_prob-old_log_prob).sum() {(log_prob-old_log_prob).sum()}")`

- Use the following script (with env variables `export N_GPUS=<number of gpus>` and `CUDA_VISIBLE_DEVICES`):
```shell
export RAY_DEDUP_LOGS=0
export HYDRA_FULL_ERROR=1

export ROLLOUT_TP_SIZE=1

export DATA_DIR=/data/tinyzero
export BASE_MODEL=/data/huggingface/hub/models--Qwen--Qwen2.5-3B-Instruct/snapshots/aa8e72537993ba99e69dfaafa59ed015b17504d1
export CHECKPOINTS_DIR=/data/tinyzero/checkpoints_$EXPERIMENT_NAME
export REWARD_FUNCTION_PATH=TinyZero/verl/utils/reward_score/countdown.py

export BATCH_SIZE=256
export VAL_BATCH_SIZE=100
export MINI_BATCH_SIZE=128
export MICRO_BATCH_SIZE=32
export N_GROUP=4

PYTHONUNBUFFERED=1 python3 -m verl.trainer.main_ppo \
    data.train_files=$DATA_DIR/train.parquet \
    data.val_files=$DATA_DIR/test.parquet \
    data.train_batch_size=$BATCH_SIZE \
    data.val_batch_size=1312 \
    data.max_prompt_length=256 \
    data.max_response_length=1024 \
    actor_rollout_ref.rollout.do_sample=true \
    actor_rollout_ref.rollout.temperature=1 \
    actor_rollout_ref.rollout.top_p=1 \
    actor_rollout_ref.rollout.n=$N_GROUP \
    actor_rollout_ref.actor.use_kl_loss=true \
    actor_rollout_ref.actor.kl_loss_type="low_var_kl" \
    actor_rollout_ref.actor.kl_loss_coef=0.005 \
    actor_rollout_ref.actor.entropy_coeff=0.001 \
    algorithm.adv_estimator=grpo \
    algorithm.use_kl_in_reward=false \
    actor_rollout_ref.model.path=$BASE_MODEL \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.use_dynamic_bsz=True \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.ppo_epochs=2 \
    actor_rollout_ref.actor.ppo_mini_batch_size=$MINI_BATCH_SIZE \
    actor_rollout_ref.actor.ppo_micro_batch_size=$MICRO_BATCH_SIZE \
    actor_rollout_ref.rollout.log_prob_micro_batch_size=$MICRO_BATCH_SIZE \
    actor_rollout_ref.ref.log_prob_micro_batch_size=$MICRO_BATCH_SIZE \
    actor_rollout_ref.rollout.tensor_model_parallel_size=$ROLLOUT_TP_SIZE \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.model.enable_gradient_checkpointing=true \
    actor_rollout_ref.actor.fsdp_config.param_offload=true \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=true \
    reward_model.reward_manager='naive' \
    reward_model.launch_reward_fn_async=true \
    custom_reward_function.path=$REWARD_FUNCTION_PATH \
    custom_reward_function.name="compute_score" \
    trainer.logger=['console'] \
    trainer.val_before_train=false \
    trainer.n_gpus_per_node=$N_GPUS \
    trainer.nnodes=1 \
    trainer.save_freq=100 \
    trainer.test_freq=25 \
    trainer.default_local_dir=$CHECKPOINTS_DIR \
    trainer.default_hdfs_dir=null \
    trainer.total_epochs=15
```

Let me know if you reproduced this issue and if you found the origine of it! Thank you.

**Comment by zyf8818** [comment]
Do you have solve this problem? I have met that the negative_approx_kl always be 0, however we set π not the same with π_old.

**Comment by Kirrito-k423** [comment]
定时清理检查发现：该 issue 最近一次非当前运行账号（Kirrito-k423）的公开回复或公开引用是 2025-07-11T10:27:04Z（actor=zyf8818, type=commented），截至 2026-07-09 已超过 6 个月。按照社区 issue 定时清理标准，本次先关闭该 issue。

Scheduled maintenance check: the latest public reply or public reference from an account other than the current runner (Kirrito-k423) was at 2025-07-11T10:27:04Z (actor=zyf8818, type=commented), which is more than 6 months before 2026-07-09. Under the scheduled issue cleanup rule, this issue is being closed for now.

证据 / Evidence: latest_non_self_reply_or_reference_at=2025-07-11T10:27:04Z; comments_checked=1; timeline_public_events_checked=1; issue=https://github.com/verl-project/verl/issues/2305; timeline API=https://api.github.com/repos/verl-project/verl/issues/2305/timeline.

如果作者或社区成员仍能提供新的复现、证据或需求背景，欢迎继续评论请求重开，或创建新的 issue 并引用这里的上下文。

If the author or community can still provide a fresh reproduction, evidence, or requirement context, please comment to request reopening or open a new issue and link back to this context.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
